File: LeetCode/Graph/DFS/Python/Leetcode241-different-ways-to-add-parentheses.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        res,n = [],len(expression)
        for i in range(n):
            if not expression[i].isdigit():
                left = self.diffWaysToCompute(expression[:i])
                right = self.diffWaysToCompute(expression[i + 1:])
                for lhs in left:
                    for rhs in right:
                        if expression[i] == '+':
                            res.append(lhs + rhs)
                        elif expression[i] == '-':
                            res.append(lhs - rhs)
                        else:
                            res.append(lhs * rhs)
        if not res:
            res.append(int(expression))
        return res

File: LeetCode/Graph/DFS/Rust/Leetcode241-different-ways-to-add-parentheses.rs
impl Solution {
  pub fn diff_ways_to_compute(expression: String) -> Vec<i32> {
    let (mut res, n) = (vec![], expression.len());
    if n == 0 {
      return res;
    }
    let chars = expression.as_bytes();
    for i in 0..n {
      match chars[i] {
        b'+' | b'-' | b'*' => {
          let left = Self::diff_ways_to_compute(expression[..i].to_owned());
          let right = Self::diff_ways_to_compute(expression[i + 1..].to_owned());
          for lhs in &left {
            for rhs in &right {
              match chars[i] {
                b'+' => res.push(lhs + rhs),
                b'-' => res.push(lhs - rhs),
                _ => res.push(lhs * rhs),
              }
            }
          }
        }
        _ => continue,
      }
    }
    if res.is_empty() {
      res.push(expression.parse::<i32>().unwrap());
    }
    res
  }
}
